### PR TITLE
Enable background sync

### PR DESCRIPTION
**Pull request by MasterWanna**
This is pull request for issue #1168

### CODE REVIEW

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * Enabling Set backgroundSync to every 15 minutes, and wifiOnly to true
+     * @param database no use just implement
+     */
+    @Override
+    public void migrate(@NonNull SupportSQLiteDatabase database) {
+        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
+        SharedPreferences.Editor editor = sharedPreferences.edit();
+        editor.putBoolean("wifiOnly", true);
+        editor.putString("backgroundSync", "15_minutes");
```
Comment (stefan-niedermann) [inline app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:31]: We should not only set the `15_minutes` string, but instead
1. Check if there is a preference
2. If there is a preference, remove it
3. Set the preference again, but this time as a `boolean` (`true`)

We should move away from the `String` value and go for only `on` / `off` in the future.

File: app/src/main/res/xml/preferences.xml [hunk 2]
```diff
@@ -11,10 +11,11 @@
             android:icon="@drawable/ic_network_wifi_grey600_24dp"
             android:key="@string/pref_key_wifi_only"
             android:layout="@layout/item_pref"
-            android:title="@string/settings_wifi_only" />
+            android:title="@string/settings_wifi_only"
+            app:defaultValue="true" />
```
Comment (stefan-niedermann) [inline app/src/main/res/xml/preferences.xml:15]: Why did you change the default value for this setting? It should not be touched for fixing the issue.

File: app/src/main/res/xml/preferences.xml [hunk 3]
```diff
@@ -11,10 +11,11 @@
...
 
         <ListPreference
-            android:defaultValue="@string/pref_value_sync_off"
+            android:defaultValue="@string/pref_value_sync_15_minutes"
```
Comment (stefan-niedermann) [inline app/src/main/res/xml/preferences.xml:18]: default value should not be one of thr strings in the future, but simply `true` - a boolean. Since we used a string in older app versions, we need the migration 🙂

**Review by stefan-niedermann — CHANGES_REQUESTED** [review]
Good start. Do you @MasterWanna or @Cui-Yusong have Telegram or Skype messenger by any chance? (I guess you are classmates), maybe we can exchange a bit easier this way?

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+import android.content.Context;
+import android.content.SharedPreferences;
+
+import androidx.annotation.NonNull;
+import androidx.preference.PreferenceManager;
+import androidx.room.migration.Migration;
+import androidx.sqlite.db.SupportSQLiteDatabase;
+
+// CS304 issue link : https://github.com/stefan-niedermann/nextcloud-notes/issues/1168
```
[the remark this answers is not in the document]
Comment (MasterWanna) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:11]: @Cui-Yusong has one, and I will sign up for one then.

[the remark this answers is not in the document]
Comment (Cui-Yusong) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:11]: Yes, I have a telegram account. How could we contact using telegram?

[the remark this answers is not in the document]
Comment (Cui-Yusong) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:11]: Should we create a group chat?

[the remark this answers is not in the document]
Comment (stefan-niedermann) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:11]: Open `https://t.me/` + `stefan` + <kbd>_</kbd> + `niedermann`

[the remark this answers is not in the document]
Comment (Cui-Yusong) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java]: got it

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+     * https://github.com/stefan-niedermann/nextcloud-deck/issues/531
+     */
+    @Override
+    public void migrate(@NonNull SupportSQLiteDatabase database) {
+        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
+        SharedPreferences.Editor editor = sharedPreferences.edit();
+        if (sharedPreferences.contains("backgroundSync")) {
+            editor.remove("backgroundSync");
+        }
+        editor.putBoolean("backgroundSync", true);
```
Comment (stefan-niedermann) [inline app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:30]: The migration should check the current value:
- If it is set to `off`, then use `false` as new boolean value
- If it does not exist, don't explicitly add it (the `defaultValue` will take care of it)
- If it exists and is any other value than `off`, set the preference explicitly to `true` just like you did it here

**Review by stefan-niedermann — CHANGES_REQUESTED** [review]

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java (hunk 5 above)
Comment (MasterWanna) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:30]: Yes, it should be considered.

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/SyncWorker.java [hunk 6]
```diff
@@ -44,7 +43,14 @@ public Result doWork() {
...
 
-    public static void update(@NonNull Context context, Boolean backgroundSync) {
+    /**
+     * Set up sync work to enabled every 15 minutes or just disabled
+     * https://github.com/stefan-niedermann/nextcloud-notes/issues/1168
+     * @param context the application
+     * @param backgroundSync the toggle result backgroundSync
+     */
+
+    public static void update(@NonNull Context context, @NonNull Boolean backgroundSync) {
```
Comment (stefan-niedermann) [inline app/src/main/java/it/niedermann/owncloud/notes/persistence/SyncWorker.java:53]: Instead of a `@NonNull Boolean`, use just a `boolean` which is basically the same 😉

**Review by stefan-niedermann — CHANGES_REQUESTED** [review]

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/SyncWorker.java (hunk 6 above)
Comment (MasterWanna) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/SyncWorker.java]: Yes, they are the same. I have a question that are there some special cases causes them to be different?

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (MasterWanna) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java]: Hi, it seems that you did not add me, so I cannot send messages.

[the remark this answers is not in the document]
Comment (Cui-Yusong) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java]: My telegram account is +86 159-0543-1169 , could you please add me? @stefan-niedermann 
I find some problems when adding you .... Thanks~ 😄

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/SyncWorker.java (hunk 6 above)
Comment (stefan-niedermann) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/SyncWorker.java]: The difference is, that a `@NonNull Boolean` **can** be `null`, it is just not *supposed* to be `null`. IDEs like Android Studio will be able to display warnings if you pass `null` into this method, but it will not fail if you actually do it. If you use a value from a unknown source, the annotation will not be able to ensure that there is no `null` passed, while the application directly will fail early at this moment, because the implicit unboxing will throw an exception in this case.

So for primitive values, if you want to make sure, that there is no `null`, go for the primitive type. For all other object types and `String`, go for the annotation.

Other languages like kotlin distinguish between `Boolean` and `Boolean?` as type and provide null safety on compile time, while the `@NonNull` and `@Nullable` annotations in java will not provider compile time "safety" but will support developers well, when they are used across the application. They add more semantic to the wanted type and make the assumptions of the developer more clear.

Comment (MasterWanna) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/SyncWorker.java]: Ok, thanks.

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        if (sharedPreferences.contains("backgroundSync")) {
+            if (sharedPreferences.getString("backgroundSync", "on").equals("off")) {
+                editor.remove("backgroundSync");
+                editor.putBoolean("backgroundSync", false);
+            } else {
+                editor.remove("backgroundSync");
+                editor.putBoolean("backgroundSync", true);
+            }
+        }
+        editor.apply();
```
Comment (stefan-niedermann) [inline app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:38]: We need to update the `SyncWorker` to make it aware of the "new" setting:

```java
SyncWorker.update(context, true);
```

I'd say we need to call it with `true` in case the `SharedPreferences`
- did **not** contain a setting, because the default behavior now changed
- **did** contain a setting which was **not** `off`

Otherwise the new default will only be applied on new installations and installations which change the setting after performing the update.

I recommend you to test the behavior with the various combinations (No setting, setting `off`, setting `15_minutes`, setting `6_hours`, etc...) with this approach:

1. Install the Notes app on an Android device and set it up
2. Make a snapshot of the emulator (three dots menu -> Take snapshot)
  grafik
3. Validate the old setting in the device explorer of Android Studio
  grafik
4. Install the branch version
5. Validate the new setting in the device explorer (Make sure to right click -> synchronize first, before validating)
6. Check the Logcat (filter for `SyncWorker`) to verify whether it run or not
7. Repeat for each combination (you can now easily restart from your snapshot, see point 2)

**Review by stefan-niedermann — CHANGES_REQUESTED** [review]

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java (hunk 7 above)
Comment (MasterWanna) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:38]: Is it that we need to call `SyncWorker.update` during `Migration`, otherwise the settings are not updated?

Comment (stefan-niedermann) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:38]: Yes, during the migration.

Oh, and i just saw, that we also need to call it additionally for new installations. So also call it when importing an account the first time:

https://github.com/stefan-niedermann/nextcloud-notes/blob/0348f0a642df5d1a35008f7adcd95de1f4449daf/app/src/main/java/it/niedermann/owncloud/notes/importaccount/ImportAccountActivity.java#L91 (right before `runOnUiThread(() -> {...`) with `true`.

Comment (MasterWanna) [reply app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:44]: I have updated in migration, and in `ImportAccountActivity.java`, I dont know whether my operation is correct or not.

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        super(21, 22);
+        this.context = context;
+    }
+
+    @Override
+    public void migrate(@NonNull SupportSQLiteDatabase database) {
+        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
+        SharedPreferences.Editor editor = sharedPreferences.edit();
+        if (sharedPreferences.contains("backgroundSync")) {
+            if (sharedPreferences.getString("backgroundSync", "on").equals("off")) {
```
Comment (stefan-niedermann) [inline app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:32]: Please use `""` as default value instead of `"on"`. The default value should never be used anyway, because we already check with `.contains(...)` the line before, but we never used the value `"on"` and it will be confusing to read it here in the future.

File: app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+                editor.remove("backgroundSync");
+                editor.putBoolean("backgroundSync", false);
+                SyncWorker.update(context, false);
```
Comment (stefan-niedermann) [inline app/src/main/java/it/niedermann/owncloud/notes/persistence/migration/Migration_21_22.java:35]: No need to explicitly call it with `false` in this case, because if the background sync was explicitly switched to `off`, it is already disabled and the behavior will not change while migrating - just remove this line.

File: app/src/main/java/it/niedermann/owncloud/notes/preferences/PreferencesFragment.java [hunk 10]
```diff
@@ -31,12 +29,18 @@
...
     @Override
     public void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
     }
 
+    /**
+     * Turn background sync from from {@link String} values to {@link Boolean} values
+     * https://github.com/stefan-niedermann/nextcloud-notes/issues/1168
+     */
+
```
Comment (stefan-niedermann) [inline app/src/main/java/it/niedermann/owncloud/notes/preferences/PreferencesFragment.java:43]: Remove empty line

File: app/src/main/java/it/niedermann/owncloud/notes/preferences/PreferencesFragment.java [hunk 11]
```diff
@@ -31,12 +29,18 @@
...
 
     @Override
     public void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
     }
 
+    /**
+     * Turn background sync from from {@link String} values to {@link Boolean} values
+     * https://github.com/stefan-niedermann/nextcloud-notes/issues/1168
+     */
```
Comment (stefan-niedermann) [inline app/src/main/java/it/niedermann/owncloud/notes/preferences/PreferencesFragment.java:42]: While this description might fit to the Migration, it does not provide any value to this method. Please remove it or merge it with the description in the Migration.

**Review by stefan-niedermann — CHANGES_REQUESTED** [review]

File: app/src/main/java/it/niedermann/owncloud/notes/importaccount/ImportAccountActivity.java [hunk 12]
```diff
@@ -90,8 +94,17 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+                            /**
+                             * Update syncing when adding account
+                             * https://github.com/stefan-niedermann/nextcloud-deck/issues/531
+                             * @param account the account to add
+                             */
                             @Override
                             public void onSuccess(Account account) {
+                                Context context = getApplicationContext();
+                                SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
+                                SyncWorker.update(context, !sharedPreferences.getString("backgroundSync", "on").equals("off"));
```
[the remark this answers is not in the document]
Comment (MasterWanna) [reply app/src/main/java/it/niedermann/owncloud/notes/importaccount/ImportAccountActivity.java:107]: I forgot about this is different to that in `Migration`, this value type should be different to previous value type.

**Comment by Cui-Yusong** [comment]
Hi, I wonder if there's any issues need to be fixed right now? Here's a member of my team feel like to make further knowledge of nextcloud-notes. Thanks a lot！

**Comment by MasterWanna** [comment]
Hi, is there any more I need to improve, or is is Ok in current version?

**Comment by stefan-niedermann** [comment]
No it looks good codewise. Just give me a few days to test each possible combination to make sure no scenario breaks.

I am sorry for the delay, but i am doing this only in my spare time 🙂

**Review by stefan-niedermann — APPROVED** [review]

**Comment by MasterWanna** [comment]
Ok, I got it, thanks!

**Comment by stefan-niedermann** [comment]
Good job! Merged!
